src/helpers/pull-request-parsing.ts

```diff
@@ -0,0 +1,104 @@
+import { encode } from "gpt-tokenizer";
+import { TokenLimits } from "../types/llm";
+import { logger } from "./errors";
+import { EncodeOptions } from "gpt-tokenizer/esm/GptEncoding";
+
+export async function processPullRequestDiff(diff: string, tokenLimits: TokenLimits) {
+  const { runningTokenCount, tokensRemaining } = tokenLimits;
+
+  // parse the diff into per-file diffs for quicker processing
+  const perFileDiffs = parsePerFileDiffs(diff);
+
+  // quick estimate using a simple heuristic; 3.5 characters per token
+  const estimatedFileDiffStats = perFileDiffs.map(({ filename, diffContent }) => {
```

Comment:
Actually we essentially have unlimited compute but we have many restrictions on api calls/rate limits so I think it makes sense to optimize for minimal api calls if that's an option.

Reply:
You should run it and QA it properly but files are not fetched. The single PR diff is split into sections based on the file sections included within the raw diff as per the regex used to split the diff.

> I don't think it's worth saving one or more Octokit calls at the expense of increased memory usage and processing time.

It's an action, we have 6 hours and no cap I'm aware of in terms of memory, but this processes things fast, which was my intention considering how slow it can be to encode thousands of lines of diff, I know this by experimenting with various methods.

> A better approach would be to compare the commit hashes and check the file sizes between them, fetching files selectively instead of the current method.

That raises the API calls and needs promise.all refactoring. My approach literally uses just the diff of the PR. 

Why fetch any files at all, we only send the diff not the full file, the diff is what's been changed within the file not the whole thing so why would we use the full file sizes that also raising api calls in the process and needs additional math logic to calculate differences etc - It makes sense to just use the diff.

Reply:
Maybe it's best I close this PR

1. It involves quite a lot of changes that may or may not clash with your refactoring and vision for how things should work, as I appear to be stepping wrong no matter what it seems.
2. I feel my opinions and work is repeatedly being dismissed without proper QA or review and it's feeling combative, and I'm not vying for ownership of this plugin. I did write and rewrite it three times for V1 and V2 so I felt a sense of ownership to it but I'm happy to step back and get out of your way with things. I can focus on other areas of development no problem.
3. I'm happy to help if requested but I think you and 0x4007 have got things more than covered.

Please feel free to close the PR or take over it again, no hard feelings or anything, I just don't want any tension with anyone in the org that's all.